src/Algebra/Morphism/Bundles.agda

```diff
@@ -0,0 +1,223 @@
+-- Morphisms between Semirings
+------------------------------------------------------------------------
+
+record SemiringHomomorphism (A : Semiring a ℓa) (B : Semiring b ℓb) : Set (a ⊔ b ⊔ ℓa ⊔ ℓb) where
+  private
+    module A = Semiring A
+    module B = Semiring B
+
+  field
+    ⟦_⟧ : A.Carrier → B.Carrier
+
+    isSemiringHomomorphism : IsSemiringHomomorphism A.rawSemiring B.rawSemiring ⟦_⟧
+
+  open IsSemiringHomomorphism isSemiringHomomorphism public
+
+  nearSemiringHomomorphism : NearSemiringHomomorphism A.nearSemiring B.nearSemiring
+  nearSemiringHomomorphism = record { isNearSemiringHomomorphism = isNearSemiringHomomorphism }
+
+  open NearSemiringHomomorphism nearSemiringHomomorphism public
+    using (*-magmaHomomorphism; +-monoidHomomorphism)
+
+  *-monoidHomomorphism : MonoidHomomorphism A.*-monoid B.*-monoid
+  *-monoidHomomorphism = record { isMonoidHomomorphism = *-isMonoidHomomorphism }
+
+------------------------------------------------------------------------
+-- Morphisms between KleeneAlgebras
+------------------------------------------------------------------------
+
+record KleeneAlgebraHomomorphism (A : KleeneAlgebra a ℓa) (B : KleeneAlgebra b ℓb) : Set (a ⊔ b ⊔ ℓa ⊔ ℓb) where
+  private
+    module A = KleeneAlgebra A
+    module B = KleeneAlgebra B
+
+  field
+    ⟦_⟧ : A.Carrier → B.Carrier
+    isKleeneAlgebraHomomorphism : IsKleeneAlgebraHomomorphism A.rawKleeneAlgebra B.rawKleeneAlgebra ⟦_⟧
+
+  open IsKleeneAlgebraHomomorphism isKleeneAlgebraHomomorphism public
+
+  semiringHomomorphism : SemiringHomomorphism A.semiring B.semiring
+  semiringHomomorphism = record { isSemiringHomomorphism = isSemiringHomomorphism }
+
+  open SemiringHomomorphism semiringHomomorphism public
+    hiding (*-isMagmaHomomorphism; *-isMonoidHomomorphism)
+
+------------------------------------------------------------------------
+-- Morphisms between RingWithoutOnes
+------------------------------------------------------------------------
+
+record RingWithoutOneHomomorphism (A : RingWithoutOne a ℓa) (B : RingWithoutOne b ℓb) : Set (a ⊔ b ⊔ ℓa ⊔ ℓb) where
+  private
+    module A = RingWithoutOne A
+    module B = RingWithoutOne B
+
+  field
+    ⟦_⟧ : A.Carrier → B.Carrier
+    isRingWithoutOneHomomorphism : IsRingWithoutOneHomomorphism A.rawRingWithoutOne B.rawRingWithoutOne ⟦_⟧
+
+  open IsRingWithoutOneHomomorphism isRingWithoutOneHomomorphism public
+
+  +-groupHomomorphism : GroupHomomorphism A.+-group B.+-group
+  +-groupHomomorphism = record { isGroupHomomorphism = +-isGroupHomomorphism }
+
+  *-magmaHomomorphism : MagmaHomomorphism A.*-magma B.*-magma
+  *-magmaHomomorphism = record { isMagmaHomomorphism = *-isMagmaHomomorphism }
+
+------------------------------------------------------------------------
+-- Morphisms between Rings
+------------------------------------------------------------------------
+
+record RingHomomorphism (A : Ring a ℓa) (B : Ring b ℓb) : Set (a ⊔ b ⊔ ℓa ⊔ ℓb) where
```

Comment:
This should expose
* `semiringHomomorphism`
* `ringWithoutOneHomomorphism`
* `nearSemiringHomomorphism`
* `+-magma-homomorphism`
* `+-monoid-homomorphsim`
* `*-magma-homomorphism`

Reply:
Taking a deep breath after `RingWithoutOneHomomorphism`...

Reply:
... and `*-monoidHomomorphism`?

Reply:
If it's not there already (I think I convinced myself that it was earlier)

Reply:
OK, I think I managed to do all of those... by filling in some more diamonds... cf. #1688 and friends #2254 etc. which might want to be revisited after these additions...?